File: plan1/1.py
'''
Author: He Weijie 129854778+IntronRewrite@users.noreply.github.com
Date: 2024-11-24 14:21:57
LastEditors: IntronRewrite weijiehe@sdust.edu.com
LastEditTime: 2024-11-25 03:34:05
FilePath: \IndustrialVision\plan1\1.py
Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
'''
import time
import open3d as o3d
import numpy as np
import matplotlib.pyplot as plt


def compute_average_distance(pcd):
    distances = pcd.compute_nearest_neighbor_distance()
    avg_dist = np.mean(distances)
    return avg_dist


pcd_src = o3d.io.read_point_cloud(r'..\src_model.ply')
print('读取src点云数据成功')
pcd_ref = o3d.io.read_point_cloud(r'..\ref_model.ply')
print('读取ref点云数据成功')

start_time = time.time()
#对src聚类
# avg_dist = compute_average_distance(pcd_src)
avg_dist = 1.8126511386464113
print(f"点云的平均距离: {avg_dist}")

# 去除pcd_src的离群点
# 欧式聚类pcd_src
with o3d.utility.VerbosityContextManager(o3d.utility.VerbosityLevel.Debug) as cm:
    labels = np.array(pcd_src.cluster_dbscan(eps=avg_dist*4.2, min_points=1, print_progress=True))
max_label = labels.max()
print(f"point cloud has {max_label + 1} clusters")

# 为每个聚类分配颜色
colors = plt.get_cmap("tab20")(labels / (max_label if max_label > 0 else 1))
colors[labels < 0] = 0
pcd_src.colors = o3d.utility.Vector3dVector(colors[:, :3])
print('欧式聚类成功')
# # 可视化聚类结果
# o3d.visualization.draw_geometries([pcd_src])
# print('聚类结果可视化成功')


# 只保留点最多的一个聚类
largest_cluster_idx = np.argmax(np.bincount(labels[labels >= 0]))
pcd_src = pcd_src.select_by_index(np.where(labels == largest_cluster_idx)[0])
print('保留点最多的一个聚类成功')
# # 可视化点云
# o3d.visualization.draw_geometries([pcd_src])
# print('点云可视化成功')
end_time = time.time()
print('聚类时间{:.2f}秒'.format(end_time - start_time))

#粗配准
start_time = time.time()
src_oriented_bounding_box = pcd_src.get_oriented_bounding_box()
src_oriented_bounding_box.color = (0, 1, 0)
print('获取src定向边界框成功')
# o3d.visualization.draw_geometries([pcd_src, src_oriented_bounding_box])
print('src可视化成功')
ref_oriented_bounding_box = pcd_ref.get_oriented_bounding_box()
ref_oriented_bounding_box.color = (1, 0, 0)
print('获取ref定向边界框成功')
# o3d.visualization.draw_geometries([pcd_ref, ref_oriented_bounding_box])
# print('ref可视化成功')

# Set pcd_ref color to a single color (e.g., red)
pcd_ref.paint_uniform_color([1, 0, 0])

# 获取src定向边界框的中心
src_center = src_oriented_bounding_box.get_center()
# 获取ref定向边界框的中心
ref_center = ref_oriented_bounding_box.get_center()


# 将src点云和定向边界框移动到原点
pcd_src.translate(-src_center)
src_oriented_bounding_box.translate(-src_center)
# 将ref点云和定向边界框移动到原点
pcd_ref.translate(-ref_center)
ref_oriented_bounding_box.translate(-ref_center)


o3d.visualization.draw_geometries([pcd_src, src_oriented_bounding_box,pcd_ref, ref_oriented_bounding_box])
print('ref点云和定向边界框已移动到原点')
print('src点云和定向边界框已移动到原点')



# 计算src和ref定向边界框的旋转矩阵
src_rotation = src_oriented_bounding_box.R
ref_rotation = ref_oriented_bounding_box.R
print(src_rotation)
print(ref_rotation)


# 将src点云和定向边界框旋转到与ref对齐
pcd_src.rotate(ref_rotation @ np.linalg.inv(src_rotation), center=(0, 0, 0))
src_oriented_bounding_box.rotate(ref_rotation @ np.linalg.inv(src_rotation), center=(0, 0, 0))
end_time = time.time()
print('聚类时间{:.2f}秒'.format(end_time - start_time))

o3d.visualization.draw_geometries([pcd_src, src_oriented_bounding_box, pcd_ref, ref_oriented_bounding_box])
print('src点云和定向边界框已旋转对齐ref')


# 精配准src和ref点云
start_time = time.time()
threshold = 1 # 配准距离阈值
trans_init = np.eye(4)  # 初始变换矩阵

# 使用ICP算法进行精配准
reg_p2p = o3d.pipelines.registration.registration_icp(
    pcd_src, pcd_ref, threshold, trans_init,
    o3d.pipelines.registration.TransformationEstimationPointToPoint()
)

# 获取精配准后的变换矩阵
transformation_icp = reg_p2p.transformation
print("ICP 变换矩阵:")
print(transformation_icp)

# 将src点云应用精配准后的变换矩阵
pcd_src.transform(transformation_icp)
end_time = time.time()
print('聚类时间{:.2f}秒'.format(end_time - start_time))

# 可视化精配准结果
o3d.visualization.draw_geometries([pcd_src, pcd_ref])
print('src点云和ref点云精配准成功')

File: plan2/1.cpp
/*
 * @Author: IntronRewrite weijiehe@sdust.edu.com
 * @Date: 2024-11-24 22:11:21
 * @LastEditors: IntronRewrite weijiehe@sdust.edu.com
 * @LastEditTime: 2024-11-28 22:07:17
 * @FilePath: /lw/IndustrialVision/plan2/1.cpp
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
/*
 *                        _oo0oo_
 *                       o8888888o
 *                       88" . "88
 *                       (| -_- |)
 *                       0\  =  /0
 *                     ___/`---'\___
 *                   .' \\|     |// '.
 *                  / \\|||  :  |||// \
 *                 / _||||| -:- |||||- \
 *                |   | \\\  - /// |   |
 *                | \_|  ''\---/''  |_/ |
 *                \  .-\__  '-'  ___/-. /
 *              ___'. .'  /--.--\  `. .'___
 *           ."" '<  `.___\_<|>_/___.' >' "".
 *          | | :  `- \`.;`\ _ /`;.`/ - ` : | |
 *          \  \ `_.   \_ __\ /__ _/   .-` /  /
 *      =====`-.____`.___ \_____/___.-`___.-'=====
 *                        `=---='
 * 
 * 
 *      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * 
 *            佛祖保佑     永不宕机     永无BUG
 * 
 *        佛曰:  
 *                写字楼里写字间，写字间里程序员；  
 *                程序人员写程序，又拿程序换酒钱。  
 *                酒醒只在网上坐，酒醉还来网下眠；  
 *                酒醉酒醒日复日，网上网下年复年。  
 *                但愿老死电脑间，不愿鞠躬老板前；  
 *                奔驰宝马贵者趣，公交自行程序员。  
 *                别人笑我忒疯癫，我笑自己命太贱；  
 *                不见满街漂亮妹，哪个归得程序员？
 */

/*
 * @Author: IntronRewrite weijiehe@sdust.edu.com
 * @Date: 2024-11-24 22:11:21
 * @LastEditors: IntronRewrite weijiehe@sdust.edu.com
 * @LastEditTime: 2024-11-24 22:42:55
 * @FilePath: /lw/IndustrialVision/plan2/1.cpp
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
#include <pcl/io/ply_io.h>
#include <pcl/point_types.h>
#include <pcl/visualization/pcl_visualizer.h>

int main(int argc, char** argv) {
    // 创建点云对象
    pcl::PointCloud<pcl::PointXYZ>::Ptr ref_cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::PointCloud<pcl::PointXYZ>::Ptr src_cloud(new pcl::PointCloud<pcl::PointXYZ>);

    // 加载点云文件
    if (pcl::io::loadPLYFile<pcl::PointXYZ>("../ref_model.ply", *ref_cloud) == -1) {
        PCL_ERROR("无法读取文件 ref_model.ply \n");
        return (-1);
    }
    if (pcl::io::loadPLYFile<pcl::PointXYZ>("../src_model.ply", *src_cloud) == -1) {
        PCL_ERROR("无法读取文件 src_model.ply \n");
        return (-1);
    }

    // 创建 PCLVisualizer 对象
    pcl::visualization::PCLVisualizer::Ptr viewer(new pcl::visualization::PCLVisualizer("3D Viewer"));
    viewer->setBackgroundColor(1, 1, 1); // 设置背景颜色为白色

    // 将参考点云添加到查看器
    pcl::visualization::PointCloudColorHandlerCustom<pcl::PointXYZ> ref_color(ref_cloud, 255, 0, 0); // 红色
    viewer->addPointCloud<pcl::PointXYZ>(ref_cloud, ref_color, "ref cloud");

    // 将源点云添加到查看器
    pcl::visualization::PointCloudColorHandlerCustom<pcl::PointXYZ> src_color(src_cloud, 0, 255, 0); // 绿色
    viewer->addPointCloud<pcl::PointXYZ>(src_cloud, src_color, "src cloud");

    // 设置点云属性
    viewer->setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 1, "ref cloud");
    viewer->setPointCloudRenderingProperties(pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 1, "src cloud");

    // 添加坐标系
    viewer->addCoordinateSystem(1.0);
    viewer->initCameraParameters(); // 初始化相机参数

    // 主循环
    while (!viewer->wasStopped()) {
        viewer->spinOnce(1000000); // 每次循环等待100毫秒
    }

    return 0;
}